### PR TITLE
[ARM][CPU] Implement LogicalNot JIT Emitter for ARM64 SIMD in OpenVino

### LINKED ISSUE #24445: [Good First Issue] [ARM]: Implement CPU plugin just-in-time emitter for LogicalNot operation
### Context

[JIT Emitters](https://github.com/openvinotoolkit/openvino/blob/42f1cb095143f19c0b9ee25836c29748bc8d9bf2/src/plugins/intel_cpu/src/emitters/README.md) are part of code generation feature (a.k.a. tensor compiler) that automatically produces highly-efficient optimized fused subgraph binary code. Each emitter implements specific operation from low level OpenVINO dialect.

### Prerequisites
Recommended to use ARM CPU based platform for development (e.g. Mac, Raspberry Pi etc). Emulators (e.g. QEMU) is still an option, but not that convenient, especially for final performance evaluation. 

### What needs to be done?

* Create `fp32` [LogicalNot](https://docs.openvino.ai/2024/documentation/openvino-ir-format/operation-sets/operation-specs/logical/logical-not-1.html) operation JIT emitter for ARM64 SIMD platform. Use OpenVINO [CPU plugin JIT emitters](https://github.com/openvinotoolkit/openvino/blob/42f1cb095143f19c0b9ee25836c29748bc8d9bf2/src/plugins/intel_cpu/src/emitters/README.md) documentation and [Arm A-profile A64 Instruction Set Architecture](https://developer.arm.com/documentation/ddi0602/2024-03/SIMD-FP-Instructions) documentation for details.
* Modify [ARM64 executor](https://github.com/openvinotoolkit/openvino/blob/4ef40f0f0e70abf04f13310f3e7a8f6784667dbf/src/plugins/intel_cpu/src/nodes/executors/aarch64/jit_eltwise.cpp#L20) and [ARM64 kernel](https://github.com/openvinotoolkit/openvino/blob/4ef40f0f0e70abf04f13310f3e7a8f6784667dbf/src/plugins/intel_cpu/src/nodes/kernels/aarch64/jit_uni_eltwise_generic.hpp#L80) (in both places: [create_eltwise_emitter](https://github.com/openvinotoolkit/openvino/blob/4ef40f0f0e70abf04f13310f3e7a8f6784667dbf/src/plugins/intel_cpu/src/nodes/kernels/aarch64/jit_uni_eltwise_generic.cpp#L601) and [get_supported_precisions](https://github.com/openvinotoolkit/openvino/blob/4ef40f0f0e70abf04f13310f3e7a8f6784667dbf/src/plugins/intel_cpu/src/nodes/kernels/aarch64/jit_uni_eltwise_generic.cpp#L766)) to apply developed JIT emitter.

Before emitter implementation, please, modify tests to be sure that deveoped functionality is covered by test:
  * Add `jit` kernel check in [element-wise](https://github.com/openvinotoolkit/openvino/blob/4ef40f0f0e70abf04f13310f3e7a8f6784667dbf/src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/eltwise.cpp#L257) or [activation](https://github.com/openvinotoolkit/openvino/blob/4ef40f0f0e70abf04f13310f3e7a8f6784667dbf/src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/activation.cpp#L155) test, the test must fail.
  * If test doesn't fail then add supported operation in [element-wise test instantiation](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/eltwise.cpp) or [activation test instantiation](https://github.com/openvinotoolkit/openvino/blob/4ef40f0f0e70abf04f13310f3e7a8f6784667dbf/src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/activation.cpp) if it's absent.

### Tests
Tests are disabled in default build, so ensure to add `-DENABLE_TESTS=ON` into cmake command.

[GoogleTest](https://github.com/google/googletest) is used for testing. CPU functional test target is [ov_cpu_func_tests](https://github.com/openvinotoolkit/openvino/tree/master/src/plugins/intel_cpu/tests/functional). You can use two `GoogleTest` filters for element-wise and activation operations:
* for element-wise operations:
```sh
./bin/[platform]/[build_type]/ov_cpu_func_tests --gtest_filter="*smoke*Eltwise*"
```
* for activations:
```sh
./bin/[platform]/[build_type]/ov_cpu_func_tests --gtest_filter="*smoke*Activation*"
```

### Example Pull Requests

* [Abs](https://docs.openvino.ai/2023.3/openvino_docs_ops_arithmetic_Abs_1.html) operation: https://github.com/openvinotoolkit/openvino/pull/23692
* [Sigmoid](https://docs.openvino.ai/2023.3/openvino_docs_ops_activation_Sigmoid_1.html) operation, which reuses [Exp emitter](https://github.com/openvinotoolkit/openvino/blob/4ef40f0f0e70abf04f13310f3e7a8f6784667dbf/src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp#L820): https://github.com/openvinotoolkit/openvino/pull/23065

### Resources

- [Contribution guide - start here!](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [CPU plugin JIT emitters](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/src/emitters/README.md)
- [Blog post on contributing to OpenVINO](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [User documentation](https://docs.openvino.ai/)
- [Intel DevHub Discord channel](https://discord.gg/7pVRxUwdWG) - engage in discussions, ask questions and talk to OpenVINO developers
- [How to link your Pull Request to an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue-using-the-pull-request-sidebar)

### Contact points

@eshoguli, @dmitry-gorokhov

### Ticket

CVS-137699

### PR DESCRIPTION
**Pull request by jvr0123**
### Details:
 - *Added LogicalNot JIT Emitter*

### Tickets:
 - *Closes #24445  ([Implement CPU plugin just-in-time emitter for LogicalNot operation](https://github.com/openvinotoolkit/openvino/issues/24445))*

### CODE REVIEW

**Comment by mlukasze** [comment]
hey @jvr0123 there is a numbers of conflicts to solve, before we will trigger CIs, could you, please?

**Comment by jvr0123** [comment]
> hey @jvr0123 there is a numbers of conflicts to solve, before we will trigger CIs, could you, please?

Yes, apologies for the delay I was on travel, CI should be able to trigger now

**Review by eshoguli — COMMENTED** [review]
Please, rebase to the latest master, thanks!

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 1]
```diff
@@ -803,7 +803,6 @@ std::set<std::vector<element::Type>> jit_hswish_emitter::get_supported_precision
 }
 
 /// IS_INF ///
-
```
Comment (eshoguli) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:806]: here and below: please, dont update not related part of the source code

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 2]
```diff
@@ -901,6 +900,47 @@ void jit_is_inf_emitter::register_table_entries() {
...
+    OV_CPU_JIT_EMITTER_ASSERT(exec_prc_ == ov::element::f32, "unsupported precision: " + exec_prc_.to_string());
+
+    using TReg = typename dnnl::impl::cpu::aarch64::cpu_isa_traits<isa>::TReg;
+    TReg src = TReg(in_vec_idxs[0]);
+    TReg dst = TReg(out_vec_idxs[0]);
+    TReg tmp = TReg(aux_vec_idxs[0]);
+
+    h->movi(tmp.b16, 1);
+    h->bit(src.b16, tmp.b16, src.b16);
+    h->eor(dst.b16, tmp.b16, src.b16);
```
Comment (eshoguli) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:937]: `h->not_(dst.b16, src.b16);` ?

File: src/plugins/intel_cpu/src/nodes/kernels/aarch64/jit_uni_eltwise_generic.cpp [hunk 3]
```diff
@@ -629,6 +629,15 @@ struct EltwiseEmitter<jit_is_inf_emitter> {
...
 
+template<>
+struct EltwiseEmitter<jit_logicalnot_emitter> {
+    void operator()(EltwiseEmitterContext& ctx) {
+        ctx.emitter = std::make_shared<jit_logicalnot_emitter>(ctx.host,
+                                                                 ctx.host_isa,
+                                                                 ctx.exec_prc);
+    }
+};
+
```
Comment (eshoguli) [inline src/plugins/intel_cpu/src/nodes/kernels/aarch64/jit_uni_eltwise_generic.cpp:640]: you don't need template specialization for `LogicalNot` emitter

**Comment by mlukasze** [comment]
hey @jvr0123 will you have a time to work on this PR?

**Comment by jvr0123** [comment]
> hey @jvr0123 will you have a time to work on this PR?

Yes, apologies, I'll have an update after this weekend

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp (hunk 1 above)
Comment (jvr0123) [reply src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:873]: I'm not quite sure why, but for some reason this and several other files are being changed in a similar way when the clang format target is enabled in cmake

**Comment by dmitry-gorokhov** [comment]
build_jenkins

**Comment by jvr0123** [comment]
Could you trigger CI again when you get the chance?

**Comment by dmitry-gorokhov** [comment]
build_jenkins

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 4]
```diff
@@ -1031,21 +1100,87 @@ void jit_is_nan_emitter::register_table_entries() {
...
+    using TReg = typename dnnl::impl::cpu::aarch64::cpu_isa_traits<isa>::TReg;
+    TReg src = TReg(in_vec_idxs[0]);
+    TReg dst = TReg(out_vec_idxs[0]);
+    TReg tmp1 = TReg(aux_vec_idxs[0]);
+    TReg tmp2 = TReg(aux_vec_idxs[1]);
+
+    h->eor(tmp1.b16, tmp1.b16, tmp1.b16);
+    h->fcmeq(tmp1.s, tmp1.s, src.s);
+    h->ld1r(tmp2.s, table_val2("one"));
+    h->eor(dst.b16, dst.b16, dst.b16);
```
Comment (eshoguli) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:1153]: don't need this line: `h->eor(dst.b16, dst.b16, dst.b16);`

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 5]
```diff
@@ -1031,21 +1100,87 @@ void jit_is_nan_emitter::register_table_entries() {
...
+
+    using TReg = typename dnnl::impl::cpu::aarch64::cpu_isa_traits<isa>::TReg;
+    TReg src = TReg(in_vec_idxs[0]);
+    TReg dst = TReg(out_vec_idxs[0]);
+    TReg tmp1 = TReg(aux_vec_idxs[0]);
+    TReg tmp2 = TReg(aux_vec_idxs[1]);
+
+    h->eor(tmp1.b16, tmp1.b16, tmp1.b16);
+    h->fcmeq(tmp1.s, tmp1.s, src.s);
+    h->ld1r(tmp2.s, table_val2("one"));
```
Comment (eshoguli) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:1152]: You don't need `tmp2`. You can use `dst` instead `tmp2`.

**Review by eshoguli — COMMENTED** [review]
1) we need only `0` and `1` values in input tensor (`ActivationLayerTest::generate_inputs` method)
2) please, dont update not related part of the source code, don't use `clang_format_fix_all`

**Comment by jvr0123** [comment]
Okay I applied all of the suggested changes and added the activation layer test, could you trigger CI again when you get the chance?

**Comment by dmitry-gorokhov** [comment]
build_jenkins

File: src/tests/test_utils/common_test_utils/src/node_builders/activation.cpp [hunk 6]
```diff
@@ -45,6 +45,7 @@
 #include "openvino/op/swish.hpp"
 #include "openvino/op/tan.hpp"
 #include "openvino/op/tanh.hpp"
+#include "openvino/op/logical_not.hpp"
```
Comment (dmitry-gorokhov) [inline src/tests/test_utils/common_test_utils/src/node_builders/activation.cpp:48]: Code style job failed. Please move this include upper and place after `#include "openvino/op/log.hpp"`. Clang format requires includes to be sorted.

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 7]
```diff
@@ -1031,6 +1031,68 @@ void jit_is_nan_emitter::register_table_entries() {
     push_arg_entry_of("zero", 0x00000000, true);
 }
 
+/// LOGICAL_NOT ///
+jit_logicalnot_emitter::jit_logicalnot_emitter(dnnl::impl::cpu::aarch64::jit_generator* host,
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:1035]: minor: We already have logical not emitter [implemented ](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/src/emitters/plugin/x64/jit_eltwise_emitters.cpp#L1426)for x64 architecture. I would ask you to aling naming with it: `jit_logicalnot_emitter` -> `jit_logical_not_emitter`

**Comment by dmitry-gorokhov** [comment]
build_jenkins

**Review by dmitry-gorokhov — APPROVED** [review]

**Comment by dmitry-gorokhov** [comment]
@jvr0123 I am merging the PR. 
Thank you for the contribution! Great work!

**Comment by jvr0123** [comment]
> @jvr0123 I am merging the PR. Thank you for the contribution! Great work!

Thank you! Appreciate the help and patience!

**Pull request merged**